tests/testflows/map_type/requirements/requirements.md

```diff
@@ -0,0 +1,523 @@
+```
+
+Exceptions:
+
+* when key is `NULL` the return value SHALL be `NULL`
+* when key has [Integer] type and key is either too large or negative then
+  when reading from a table column it SHALL return `zero`.
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Value.Retrieval.KeyNotFound
+version: 1.0
+
+[ClickHouse] SHALL return
+
+* zeros for [Integer] values
+* empty strings for [String] values
+* empty for [Array] values
+
+when there's no corresponding `key` defined in the `Map(key, value)` data type. 
+
+
+### Converting Tuple(Array, Array) to Map
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.TupleOfArraysToMap
+version: 1.0
+
+[ClickHouse] SHALL support converting [Tuple(Array, Array)] to `Map(key, value)` using the [CAST] function.
+
+``` sql
+SELECT CAST(([1, 2, 3], ['Ready', 'Steady', 'Go']), 'Map(UInt8, String)') AS map;
+```
+
+``` text
+┌─map───────────────────────────┐
+│ {1:'Ready',2:'Steady',3:'Go'} │
+└───────────────────────────────┘
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.TupleOfArraysMap.Invalid
+version: 1.0
+
+[ClickHouse] SHALL return an error when casting [Tuple(Array, Array)] when arguments are invalid
+such as when arrays are not of equal size.
+
+For example,
+
+```sql
+SELECT CAST(([2, 1, 1023], ['', '']), 'Map(UInt8, String)') AS map, map[10]
+```
+
+### Converting Array(Tuple(K,V)) to Map
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.ArrayOfTuplesToMap
+version: 1.0
+
+[ClickHouse] SHALL support converting [Array(Tuple(K,V))] to `Map(key, value)` using the [CAST] function.
+
+For example,
+
+```sql
+SELECT CAST(([(1,2),(3)]), 'Map(UInt8, UInt8)') AS map
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.Conversion.From.ArrayOfTuplesToMap.Invalid
+version: 1.0
+
+[ClickHouse] SHALL return an error when casting [Array(Tuple(K, V))] to `Map(key, value)`
+when [Tuple] is invalid such as when it is not a [Tuple] with two elements.
+
+For example,
+
+```sql
+SELECT CAST(([(1,2),(3)]), 'Map(UInt8, UInt8)') AS map
+```
+
+or 
+
+```sql
+SELECT CAST(([(1,2),(3,)]), 'Map(UInt8, UInt8)') AS map
+```
+
+### Keys and Values Subcolumns
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.SubColumns.Keys
+version: 1.0
+
+[ClickHouse] SHALL support `keys` subcolumn in the `Map(key, value)` type that can be used 
+to retrieve an [Array] of map keys.
+
+```sql
+SELECT m.keys FROM t_map;
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.SubColumns.Keys.ArrayFunctions
+version: 1.0
+
+[ClickHouse] SHALL support applying [Array] functions to the `keys` subcolumn in the `Map(key, value)` type.
+
+For example,
+
+```sql
+SELECT * FROM t_map WHERE has(m.keys, 'a');
+```
+
+#### RQ.SRS-018.ClickHouse.Map.DataType.SubColumns.Keys.InlineDefinedMap
+version: 1.0
+
+[ClickHouse] SHALL not support using inline defined map to get `keys` subcolumn.
```

Comment:
SHALL?

Reply:
Right now it does not support it. Again, can be changed when it supports it.

Reply:
Updated requirement to use MAY.